servicenow/stackstate_checks/servicenow/servicenow.py

```diff
@@ -0,0 +1,199 @@
+            raise ConfigurationError("Missing 'url' in instance configuration.")
+
+        return TopologyInstance(self.INSTANCE_TYPE, instance["url"])
+
+    def check(self, instance):
+        if 'url' not in instance:
+            raise ConfigurationError('ServiceNow CMDB topology instance missing "url" value.')
+        if 'user' not in instance:
+            raise ConfigurationError('ServiceNow CMDB topology instance missing "user" value.')
+        if 'password' not in instance:
+            raise ConfigurationError('ServiceNow CMDB topology instance missing "password" value.')
+
+        user = instance['user']
+        password = instance['password']
+        auth = (user, password)
+
+        base_url = instance['url']
+        batch_size = instance['batch_size']
+        instance_tags = instance.get('tags', [])
+
+        default_timeout = self.init_config.get('default_timeout', 5)
+        timeout = float(instance.get('timeout', default_timeout))
+
+        instance_config = InstanceInfo(instance_tags, base_url, auth)
+
+        relation_types = self._process_and_cache_relation_types(instance_config, timeout)
+        try:
+            self.start_snapshot()
+            self._process_components(instance_config, timeout)
+            self._process_component_relations(instance_config, batch_size, timeout, relation_types)
+        except Exception as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, message=str(e), tags=instance_tags)
+        finally:
+            self.stop_snapshot()
+
+        # Report ServiceCheck OK
+        msg = "ServiceNow CMDB instance detected at %s " % base_url
+        tags = ["url:%s" % base_url]
+        self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=tags, message=msg)
+
+    def _collect_components(self, instance_config, timeout):
+        """
+        collect components from ServiceNow CMDB's cmdb_ci table
+        (API Doc- https://developer.servicenow.com/app.do#!/rest_api_doc?v=london&id=r_TableAPI-GET)
+
+        :return: dict, raw response from CMDB
+        """
+
+        base_url = instance_config.base_url
+        auth = instance_config.auth
+        url = base_url + '/api/now/table/cmdb_ci?sysparm_fields=name,sys_id,sys_class_name,sys_created_on'
+
+        return self._get_json(url, timeout, auth)
+
+    def _process_components(self, instance_config, timeout):
+        """
+        process components fetched from CMDB
+        :return: nothing
+        """
+        instance_tags = instance_config.instance_tags
+
+        state = self._collect_components(instance_config, timeout)
+
+        for component in state['result']:
+            comp_name = component['name'].encode('utf-8')
+            external_id = str(comp_name) if isinstance(comp_name, bytes) else comp_name
+            comp_type = component['sys_class_name']
+            data = {
+                "sys_id": component['sys_id'],
```

Comment:
comp_name is only used in the externalId? What is used for the name? Can we also generate a urn based identifer for the components here? Something like: 
`urn:service-now:type:identifier` 
Is it possible that things from ServiceNow should merge with other StackPacks?

Reply:
@zandrewitte I created the identifier on basis of `sys_id` and format is `urn:servicenow:type:sys_id`..